Support multiple frequency plans on gateways

KrishnaIyer opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

Closes #761 

#### Changes
<!-- What are the changes made in this pull request? -->

- Add `frequency_plan_ids` to proto and corresponding changes to IS (by @htdvisser) 
- Add `frequency_plan_id` to `TxRequest` and use that in `ScheduleDown`.
- Move `router_config` code to `pkg/pfconfig` and support generation with multiple frequency plans. 
- Adapt code and tests.

#### Notes for Reviewers
<!--
NOTE: This section is optional.

Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->

@htdvisser : I haven't modified your commits. Please check if [](https://github.com/TheThingsNetwork/lorawan-stack/commit/696796acfae2f55252e15459cc953db22e810fab) is necessary as it was in your branch.
@rvolosatovs: Since the `router_config` stuff is very specific to the Basic Station spec, it's better if @johanstokking looks at it.

This code doesn't compile since the schedular isn't ready to accept multiple frequency plans. Though All tests pass when I hacked this part. Please target the PR for #1824 to this branch.

Once that is merged, I'll close this draft and open a PR.

#### Checklist
<!-- Make sure that this pull request is complete. -->

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, database and configuration, according to the stability commitments in `README.md`.
- [x] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

rvolosatovs commented on pkg/pfconfig/basicstationlns/basicstationlns.go:
```diff
@@ -0,0 +1,181 @@
+			return RouterConfig{}, errFrequencyPlan
+		}
+	}
+	cfg := RouterConfig{}
```
```suggestion
	conf := RouterConfig{}
```
https://github.com/TheThingsNetwork/lorawan-stack/blob/master/DEVELOPMENT.md#single-word-entities

rvolosatovs commented on pkg/pfconfig/basicstationlns/basicstationlns.go:
```diff
@@ -0,0 +1,181 @@
+	type Alias RouterConfig
+	return json.Marshal(struct {
+		Type string `json:"msgtype"`
+		Alias
```
+1 point for creativity
-2 points for clarity

Not necessary to change, but isn't there a simpler way to do this?

KrishnaIyer replied:
Hmm. I don't see one. Do you have a suggestion?

johanstokking replied:
This is fine

johanstokking commented on pkg/gatewayserver/gatewayserver.go:
```diff
@@ -381,18 +381,24 @@ func (gs *GatewayServer) Connect(ctx context.Context, frontend io.Frontend, ids
-	if err != nil {
-		return nil, err
+
+	var fps []*frequencyplans.FrequencyPlan
```
You can `make` the slice here with known capacity

johanstokking commented on pkg/gatewayserver/gatewayserver.go:
```diff
@@ -612,21 +618,29 @@ func (gs *GatewayServer) GetFrequencyPlan(ctx context.Context, ids ttnpb.Gateway
 		return nil, err
 	}
-	return gs.FrequencyPlans.GetByID(fpID)
+	var fps []*frequencyplans.FrequencyPlan
```
Idem

johanstokking commented on pkg/gatewayserver/io/basicstationlns/basicstationlns.go:
```diff
@@ -375,7 +376,8 @@ func (s *srv) handleTraffic(c echo.Context) (err error) {
 					return nil
 				}
-				up, err := jreq.ToUplinkMessage(ids, fp.BandID, receivedAt)
+				// According to spec, all frequency plans must be from the same band, so choosing the first one.
```
We can and must enforce this everywhere, at IS level. cc @htdvisser

johanstokking replied:
So this ain't true anymore

johanstokking commented on pkg/gatewayserver/io/io.go:
```diff
@@ -263,7 +265,19 @@ func (c *Connection) ScheduleDown(path *ttnpb.DownlinkPath, msg *ttnpb.DownlinkM
+			break
+		}
+	}
+	if fpIndex == len(c.fps) {
```
Unclear and too much heuristic of how indexer of for-range works. Just assign `gtwFP` in the loop and check here if it's assigned

johanstokking commented on pkg/gatewayserver/io/io.go:
```diff
@@ -245,6 +245,8 @@ func (c *Connection) SendDown(msg *ttnpb.DownlinkMessage) error {
 	return nil
 }
 
+var errUnconfiguredFrequencyPlan = errors.DefineInvalidArgument("unconfigured_frequency_plan", "frequency plan `{name}` is not configured for this gateway")
```
`errFrequencyPlanNotConfigured`

johanstokking commented on pkg/identityserver/application_registry.go:
```diff
@@ -169,16 +169,18 @@ func (is *IdentityServer) listApplications(ctx context.Context, req *ttnpb.ListA
 		return nil, err
 	}
+
+	for i, app := range apps.Applications {
```
How is this change related?

htdvisser replied:
Consistency. If we move the same logic out of the transaction for Gateways, we should also do it in the other places.

htdvisser commented on pkg/identityserver/gateway_registry.go:
```diff
@@ -57,6 +60,28 @@ func (is *IdentityServer) createGateway(ctx context.Context, req *ttnpb.CreateGa
+				return nil, errFrequencyPlans
+			}
+		}
+	}
```
I'm not sure the Identity Server should do this kind of validation. Doing that would prevent us from having (test) clusters that support experimental frequency plans.

If we do want to do this, I would write it as:

```go
if len(req.FrequencyPlanIDs) > 0 {
	fp0, err := is.FrequencyPlans.GetByID(req.FrequencyPlanIDs[0])
	if err != nil {
		return nil, err
	}
	for i := 1; i < len(req.FrequencyPlanIDs); i++ {
		fpn, err := is.FrequencyPlans.GetByID(req.FrequencyPlanIDs[i])
		if err != nil {
			return nil, err
		}
		if fpn.BandID != fp0.BandID {
			return nil, errFrequencyPlans
		}
	}
}
```

instead of fetching two frequency plans in every iteration of the loop

KrishnaIyer replied:
@johanstokking ?

johanstokking replied:
I don't think we should be enforcing this in ER indeed; it doesn't have any authority in frequency plans.

KrishnaIyer replied:
Ok? Then did I mis understand your comment here:https://github.com/TheThingsNetwork/lorawan-stack/pull/1826#discussion_r364665915?

johanstokking commented:
@KrishnaIyer what is the scope for review now? This PR is draft and CI is still failing. Let me know what to review and what the plan is with merging this

KrishnaIyer commented:
- I've fixed the failing tests (I think). 
- I've moved the check for Bands from IS to GS (while connection).
- Apart from this, can you check if the scheduler needs anything?  As we discussed, since we enforce that all frequency plans have the same band, the scheduler currently receives the first FP.
- Also, this change would require the NS to send the `FrequencyPlanID` as part of the `TxRequest`. Ping @rvolosatovs

rvolosatovs commented:
> Also, this change would require the NS to send the FrequencyPlanID as part of the TxRequest. Ping @rvolosatovs

@KrishnaIyer please see https://github.com/TheThingsNetwork/lorawan-stack/pull/1826/commits/77b81cd95468d16fa4d6c196a73ac4b4ccdc9763

KrishnaIyer commented:
> You cannot simply pass the first frequency plan to the scheduler, because some settings are taken from the frequency plan and not from the underlying band.

Yes of course not. That would be too easy. This is only to make the tests pass (There seems to be a flake in the IS).

This PR is not meant to have the changes for the scheduler. That's why [this issue](https://github.com/TheThingsNetwork/lorawan-stack/issues/1824) exists.

KrishnaIyer commented:
Opening an actual PR since #1824 is done

rvolosatovs commented:
> Opening an actual PR since #1824 is done

@KrishnaIyer you can just click "ready to review" button to convert a draft PR into an "actual PR"

This(#1826) is the "actual PR", since it contains all the discussion etc.
We're already losing a lot of context by relying heavily on GitHub and not `git` directly(e.g. long commit messages utilizing the body and not just the header) and this kind of stuff makes it even worse. Please consider closing #1864 and reopening this instead and rebasing.

KrishnaIyer commented:
That's a good point @rvolosatovs. I didn't know you could convert a draft PR to an actual one. In this case unfortunately, Github doesn't allow re-opening of Draft PRs. Even @htdvisser tried in admin mode.  So I've reopened #1864 instead. Sorry for the confusion.

rvolosatovs commented:
You cannot reopen this PR, because the commit hash does not match
You need to:
1. force-push `feature/761-multi-fplan-gtw` to 6833a20
2. reopen the PR
3. force-push `feature/761-multi-fplan-gtw` to correct commit (currently eaefe77 in #1864)

johanstokking commented on pkg/gatewayserver/io/io.go:
```diff
@@ -265,7 +269,22 @@ func (c *Connection) ScheduleDown(path *ttnpb.DownlinkPath, msg *ttnpb.DownlinkM
+		}
+	}
+	if fp == nil {
+		return 0, errFrequencyPlanNotConfigured
```
Missing attribute

johanstokking replied:
Also, we don't break API here; if there's no frequency plan in the `TxRequest`, we do what we did before

Please keep (some) test cases with the current behavior to test that

KrishnaIyer replied:
That's a good one. Updated.

johanstokking commented on pkg/gatewayserver/io/io.go:
```diff
@@ -265,7 +269,22 @@ func (c *Connection) ScheduleDown(path *ttnpb.DownlinkPath, msg *ttnpb.DownlinkM
-	phy, err := band.GetByID(c.fp.BandID)
+	var fp *frequencyplans.FrequencyPlan
+	for _, gtwFP := range c.fps {
+		reqFP, err := c.fpStore.GetByID(request.FrequencyPlanID)
```
We should not be getting the frequency plans from the store in the hot path; we have the frequency plans already loaded in the connection's gateway right?

johanstokking commented on pkg/gatewayserver/io/io.go:
```diff
@@ -265,7 +269,22 @@ func (c *Connection) ScheduleDown(path *ttnpb.DownlinkPath, msg *ttnpb.DownlinkM
+			return 0, err
+		}
+		if gtwFP.BandID == reqFP.BandID {
+			fp = gtwFP
```
Also what does this mean; the frequency plan matches if the band matches?

KrishnaIyer replied:
No indeed. The frequency plan ID should match. Fixed.

johanstokking commented on pkg/gatewayserver/scheduling/scheduler.go:
```diff
@@ -65,9 +65,14 @@ type RTTs interface {
 }
 
+var (
+	errFrequencyPlansTimeOffAir     = errors.DefineInvalidArgument("frequency_plans_timeoffair", "frequency plans must have the same time off air value")
```
nit; `time_off_air`

johanstokking commented on pkg/gatewayserver/io/io.go:
```diff
@@ -272,7 +274,22 @@ func (c *Connection) ScheduleDown(path *ttnpb.DownlinkPath, msg *ttnpb.DownlinkM
+		for _, v := range c.fps {
+			fp = v
+			break
+		}
```
Why?

KrishnaIyer replied:
https://thethingsindustries.slack.com/archives/C06UQSMPB/p1579597038008000?thread_ts=1579594672.002700&cid=C06UQSMPB

johanstokking commented on pkg/gatewayserver/scheduling/scheduler.go:
```diff
@@ -83,41 +88,75 @@ func NewScheduler(ctx context.Context, fp *frequencyplans.FrequencyPlan, enforce
+					for _, subBand := range s.subBands {
+						if subBand.IsIdentical(sb) {
+							isIdentical = true
+							break
```
`continue` and skip check on L126

KrishnaIyer replied:
Hmm why? I also want to check if there's an overlap which is on `L126`

johanstokking commented on pkg/gatewayserver/scheduling/scheduler.go:
```diff
@@ -83,41 +88,75 @@ func NewScheduler(ctx context.Context, fp *frequencyplans.FrequencyPlan, enforce
+					for _, subBand := range s.subBands {
+						if subBand.IsIdentical(sb) {
+							isIdentical = true
+							break
```
Idem

KrishnaIyer commented on pkg/gatewayserver/io/basicstationlns/basicstationlns.go:
```diff
@@ -375,6 +376,12 @@ func (s *srv) handleTraffic(c echo.Context) (err error) {
+				for _, v := range fps {
+					fp = v
+					break
+				}
```
because `fps` is a `map[string]*frequencyplans.FrequencyPlan`

johanstokking replied:
1. Use a slice
2. In this case, you don't care about the frequency plan, but about the band (L385). If the spec says all frequency plans are from the same band, why not introduce a field `BandID` and keep away from this loop?

KrishnaIyer commented:
Ok @johanstokking, I've explicitly used `BandID` where applicable.
